mantidproject/mantid
# Add reduction algorithm to refl algorithm factory

This PR adds functionality to the `ReflAlgorithmFactory` to perform reduction by running `ReflectometryReductionOneAuto`. 

The main changes are:

- Functions have been added to `RowProcessingAlgorithm.h` to convert the `PreviewRow` model to an algorithm to do the reduction. This uses a lot of the same under-the-hood functionality as some existing functions for a very similar algorithm, and the tests are also very similar. 
- Methods have been added to the `ReflAlgorithmFactory` to expose the new functions to the reflectometry GUI (although they're unused as yet).
- The existing method `makeReductionAlgorithm` has been renamed because it relates to row processing.
- The algorithm properties are now set by their correct value type rather than as string values (which had been done for historic reasons). This fixes some test failures on Ubuntu 20.04 which were failing due to float values being represented with a comma in strings instead of a period for the decimal place. It is also the more correct way of doing things.
- The tests have been refactored to make code reuse easier.
- Some TODO tests have been added for when lookup by angle is implemented (planned in a future PR)

**To test:**

The new functionality is not exposed in the GUI yet so just check that existing processing functionality still works:

- Open the ISIS Reflectometry GUI
- Go to the menu Batch->Load and load `INTER_NR_test.json` from the ISIS sample data. This should populate the table. Click the arrow next to the first group to expand it.
- Click the group and click Process. The group and child rows should all turn green to indicate they processed successfully.

Part of #34106 

*This does not require release notes* because **it only affects functionality that is not yet available in the release build**

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
